File: learn_py/unit_converter.py
valid_units = ["kilometer", "meter", "centimeter", "miles", "foot", "inch"]

def convert_unit(first_unit, second_unit, unit_value):
    conversion_factors = {"kilometer": 1000, "meter": 1, "centimeter": 0.01,
"miles": 1603.34, "foot": 0.3048, "inch": 0.0254}
    value_in_meters = unit_value * conversion_factors[first_unit]
    result = value_in_meters / conversion_factors[second_unit]
    return result

def validate_unit(in_unit):
    if in_unit not in valid_units:
        print("invalid unit")
        print(f"valid units: {valid_units}")
        return True

def get_int(first_unit):
    try:
        return float(input(f"enter the value in {first_unit}"))
    except ValueError:
        print("invalid number")

print(f"valid units: {valid_units}")
while True:
    first_unit = input("enter the unit to convert from: ").lower().strip()
    if validate_unit(first_unit):
        continue
    second_unit = input("enter the unit to convert into: ").lower().strip()
    if validate_unit(second_unit):
        continue
    unit_value = get_int(first_unit)
    print(convert_unit(first_unit, second_unit, unit_value))

File: learn_c/unit_converter.cpp
#include <iostream>
#include <limits>
#include <algorithm>
#include <cctype>
#include <set>

double get_num() {
    double temp;
    while (true) {
        std::cin >> temp;
        if (std::cin.fail()) {
            std::cout << "invalid number" << '\n';
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        } else {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            return temp;
        }
    }
}

std::string validate_input(const std::set<std::string>& valid_units) {
    std::string temp;
    while (true) {
        std::cin >> temp;
        std::transform(temp.begin(), temp.end(), temp.begin(), [](unsigned char x) {return std::tolower(x);});

        if (valid_units.find(temp) != valid_units.end()) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            return temp;
        } else {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            std::cout << "invalid unit" << '\n';
        }
    }
}

int main() {
    std::set<std::string> valid_units = {"kilometer", "meter", "centimeter", "mile", "yard", "foot", "inch"};
    double value_in, value_out;
    std::string input_unit, output_unit;
    std::cout << "valid units:" << '\n' << "kilometer, meter, centimeter, mile, yard, foot, inch" << '\n' << '\n';

    std::cout << "enter the unit to convert from: ";
    input_unit = validate_input(valid_units);
    std::cout << "enter the unit to convert into: ";
    output_unit = validate_input(valid_units);
    std::cout << "enter the value in " << input_unit << ": ";
    value_in = get_num();

    return 0;
}
